src/owl/fftpack/owl_fft_generic.mli

```diff
@@ -7,45 +7,93 @@
 
 open Owl_dense_ndarray_generic
 
-(** {5 Basic functions} *)
+(** {5 Discrete Fourier Transforms functions} *)
 
-val fft : ?axis:int -> (Complex.t, 'a) t -> (Complex.t, 'a) t
-(**
-[fft ~axis x] performs 1-dimensional FFT on a complex input. [axis] is the
-highest dimension if not specified. The return is not scaled.
- *)
+val fft
+  :  ?axis:int
+  -> ?norm:int
+  -> ?nthreads:int
+  -> (Complex.t, 'a) t
+  -> (Complex.t, 'a) t
+(** [fft ~axis x] performs 1-dimensional FFT on a complex input. [axis] is the
+    highest dimension if not specified. The return is not scaled. *)
 
-val ifft : ?axis:int -> (Complex.t, 'a) t -> (Complex.t, 'a) t
-(**
-[ifft ~axis x] performs inverse 1-dimensional FFT on a complex input. The parameter [axis]
-indicates the highest dimension by default.
- *)
+val ifft
+  :  ?axis:int
+  -> ?norm:int
+  -> ?nthreads:int
+  -> (Complex.t, 'a) t
+  -> (Complex.t, 'a) t
+(** [ifft ~axis x] performs inverse 1-dimensional FFT on a complex input. The parameter [axis]
+    indicates the highest dimension by default. *)
 
-val rfft : ?axis:int -> otyp:(Complex.t, 'a) kind -> (float, 'b) t -> (Complex.t, 'a) t
-(**
-[rfft ~axis ~otyp x] performs 1-dimensional FFT on real input along the
-[axis]. [otyp] is used to specify the output type, it must be the consistent
-precision with input [x]. You can skip this parameter by using a submodule
-with specific precision such as [Owl.Fft.S] or [Owl.Fft.D].
- *)
+val rfft
+  :  ?axis:int
+  -> ?norm:int
+  -> ?nthreads:int
+  -> otyp:('a, 'b) kind
+  -> ('c, 'd) t
+  -> ('a, 'b) t
+(** [rfft ~axis ~otyp x] performs 1-dimensional FFT on real input along the
+    [axis]. [otyp] is used to specify the output type, it must be the consistent
+    precision with input [x]. You can skip this parameter by using a submodule
+    with specific precision such as [Owl.Fft.S] or [Owl.Fft.D]. *)
 
- val irfft
+val irfft
   :  ?axis:int
   -> ?n:int
-  -> otyp:(float, 'a) kind
-  -> (Complex.t, 'b) t
-  -> (float, 'a) t
-(**
-[irfft ~axis ~n x] is the inverse function of [rfft]. Note the [n] parameter
-is used to specified the size of output.
- *)
+  -> ?norm:int
+  -> ?nthreads:int
+  -> otyp:('a, 'b) kind
+  -> ('c, 'd) t
+  -> ('a, 'b) t
+(** [irfft ~axis ~n x] is the inverse function of [rfft]. Note the [n] parameter
+    is used to specified the size of output. *)
 
 val fft2 : (Complex.t, 'a) t -> (Complex.t, 'a) t
-(**
-[fft2 x] performs 2-dimensional FFT on a complex input. The return is not scaled.
- *)
+(** [fft2 x] performs 2-dimensional FFT on a complex input. The return is not scaled. *)
 
 val ifft2 : (Complex.t, 'a) t -> (Complex.t, 'a) t
-(**
-[ifft2 x] performs inverse 2-dimensional FFT on a complex input.
- *)
+(** [ifft2 x] performs inverse 2-dimensional FFT on a complex input. *)
+
+(** {5 Discrete Cosine & Sine Transforms functions} *)
+
+val dct
+  :  ?axis:int
+  -> ?ttype:int
```

Comment:
What does this stand for? You could avoid the asserts like `assert (ttype > 0 || ttype < 5);` introducing a type for these that converts to the specific int. I am not sure if this impacts performances, but I think it can be erased by the compiler and it would enforce the constraints.

Reply:
Hi, thanks for your review. The ttype option is for the Cosine and Sine transforms type (see scipy.fft.dct and scipy.fft.dst for references as I inspired the implementation from their).

Regarding the assert, I was trying to stick with the current state of Owl. Lots of conditions checking are done with asserts through the code (which I agree with you is not optimal). Sure thing, introducing a custom type here might be beneficial.

Thanks !